docs: add deprecate/disable/removal documentation (#10144)
Homebrew/brew: docs/Deprecating-Disabling-and-Removing-Formulae.md

# Deprecating, Disabling, and Removing Formulae

There are many reasons why formulae may be deprecated, disabled, or removed. This document explains the differences between each method as well as explaining when one method should be used over another.

## Overview

This general rule of thumb can be followed:

- `deprecate!` should be used for formulae that _should_ no longer be used.
- `disable!` should be used for formulae that _cannot_ be used.
- Formulae that are no longer acceptable in homebrew/core or have been disabled for over a year should be removed.

## Deprecation

> If a user attempts to install a deprecated formula, they will be shown a warning message but the install will succeed.

Reviewer: ```suggestion
If a user attempts to install a deprecated formula, they will be shown a warning message but the install will proceed.
```
Minor nitpick; feel free to ignore. I just think that there could be other reasons why the install will fail, even if a formula isn't disabled.

Author: Makes sense to me

# Deprecating, Disabling, and Removing Formulae

There are many reasons why formulae may be deprecated, disabled, or removed. This document explains the differences between each method as well as explaining when one method should be used over another.

## Overview

This general rule of thumb can be followed:

- `deprecate!` should be used for formulae that should no longer be used
- `disable!` should be used for formulae that cannot be used
- Removal should be done with formulae that have been disabled for a long time or have a more serious issue

## Deprecation

> If a user attempts to install a disabled formula, they will be shown a warning message but the install will succeed.

Reviewer: ```suggestion
If a user attempts to install a deprecated formula, they will be shown a warning message but the install will succeed.
```

Author: Oops... guilty of copy-and-paste... 😅

```ruby
deprecate! date: "YYYY-MM-DD", because: :reason
```

The `date` parameter should be set to the date that the project became (or will become) deprecated. If there is no clear date but the formula needs to be deprecated, use today's date. If the `date` parameter is set to a date in the future, the formula will not become deprecated until that date. This can be useful if the upstream developers have indicated a date where the project will stop being supported.

The `because` parameter can be set to a preset reason (using a symbol) or a custom reason. See the [Deprecate and Disable Reasons](#deprecate-and-disable-reasons) section below for more details about the `because` parameter.

## Disabling

If a user attempts to install a disabled formula, they will be shown an error message and the install will fail.

A formula should be disabled to indicate to users that the formula cannot be used and may be removed in the future. Disabled formulae may not be able to build from source and may not have working bottles. Users who choose to attempt to install disabled formulae will likely run into issues.

> The most common reasons for disabling are when the formula cannot be built from source (meaning no bottles can be built), has been deprecated for a long time, the upstream repository has been removed, or the project has license issues that are not compatible with homebrew/core.

Reviewer: In my experience we just outright remove formulae with license issues. Should we have disabled them first?

Author: Ah, good point. I think we `disable!` if the formula has no license and remove if it's a non-oss license. I'll clarify.

**Note: disabled formulae in homebrew/core will be automatically removed one year after their disable date**

> To deprecate a formula, add a `deprecate!` call. This call should include a deprecation date (in the ISO 8601 format) and a deprecation reason:

Reviewer: ```suggestion
To disable a formula, add a `disable!` call. This call should include a deprecation date (in the ISO 8601 format) and a deprecation reason:
```

# Deprecating, Disabling, and Removing Formulae

There are many reasons why formulae may be deprecated, disabled, or removed. This document explains the differences between each method as well as explaining when one method should be used over another.

## Overview

This general rule of thumb can be followed:

- `deprecate!` should be used for formulae that _should_ no longer be used
- `disable!` should be used for formulae that _cannot_ be used

> - Formulae that have been disabled for a long time or have a more serious issue should be removed

Reviewer 2: Should we add the case of versioned formulae when we have too many and really old ones can be removed?

Reviewer 3: Right, we have a limit defined in the Versioned Formula documentation, that could also be a reason

## Deprecation

If a user attempts to install a deprecated formula, they will be shown a warning message but the install will succeed.

> A formula should be deprecated to indicate to users that the formula should not be used and may be disabled in the future. Deprecated formulae should still be able to build from source and all bottles should continue to work. Users who choose to install deprecated formulae should not have any issues.

Author: I'm hesitant to use the word `maintained` here. If whether or not the project is maintained is our standard, when would a formula be deprecated rather than disabled? I think that deprecation _is_ for unmaintained formulae because it's saying "this will stop working in the future but for now it's okay". Then, when it does stop, it's disabled.

> A formula should be deprecated to indicate to users that the formula should not be used and may be disabled in the future. Deprecated formulae should still be able to build from source and all bottles should continue to work. Users who choose to install deprecated formulae should not have any issues.

Reviewer 4: > If whether or not the project is maintained is our standard

This is our docs so I think it's fair enough.

> I think that deprecation _is_ for unmaintained formulae because it's saying "this will stop working in the future but for now it's okay".

I think we're continuing to ensure that we don't break it in CI so we're still maintaining the formula (even if upstream are not maintaining the software). Agreed that may require a further rewording.

> A formula should be deprecated to indicate to users that the formula should not be used and may be disabled in the future. Deprecated formulae should still be able to build from source and all bottles should continue to work. Users who choose to install deprecated formulae should not have any issues.

Author: Ah, so you're saying that the _formula_ is maintained by _us_, right? I agree with that.

How about, instead:

```
A formula should be deprecated to indicate to users that the formula should not be used and may be disabled in the future. Deprecated formulae should still build from source and their bottles should continue to work. These formulae should continue to receive the maintenance required to allow them to build.
```

> A formula should be deprecated to indicate to users that the formula should not be used and may be disabled in the future. Deprecated formulae should still be able to build from source and all bottles should continue to work. Users who choose to install deprecated formulae should not have any issues.

Reviewer 4: ```suggestion
A formula should be deprecated to indicate to users that the formula should not be used and will be disabled in the future. Deprecated formulae should be maintained by the Homebrew maintainers so they can still build from source and their bottles continue to work (even if unmaintained upstream). If this is not possible, they should be disabled.
```

> A formula should be deprecated to indicate to users that the formula should not be used and may be disabled in the future. Deprecated formulae should still be able to build from source and all bottles should continue to work. Users who choose to install deprecated formulae should not have any issues.

Author: I like that. Will do.

The most common reasons for deprecation are when the upstream project is deprecated, unmaintained, or archived.

To deprecate a formula, add a `deprecate!` call. This call should include a deprecation date (in the ISO 8601 format) and a deprecation reason:

```ruby
deprecate! date: "YYYY-MM-DD", because: :reason
```

> The `date` parameter should be set to the date that the project became (or will become) deprecated. If there is no clear date but the formula needs to be deprecated, use today's date. If the `date` parameter is set to a date in the future, the formula will not become deprecated until that date. This can be useful if the upstream developers have indicated a date where the project will stop being supported.

Reviewer 3: Versioned formula projects are still going, it's just the old version that got deprecated.

```suggestion
The `date` parameter should be set to the date that the formula source became (or will become) deprecated. If there is no clear date but the formula needs to be deprecated, use today's date. If the `date` parameter is set to a date in the future, the formula will not become deprecated until that date. This can be useful if the upstream developers have indicated a date where the project will stop being supported.
```

The most common reasons for disabling are when the formula cannot be built from source (meaning no bottles can be built), has been deprecated for a long time, the upstream repository has been removed, or the project has no license.

**Note: disabled formulae in homebrew/core will be automatically removed one year after their disable date**

To disable a formula, add a `disable!` call. This call should include a deprecation date (in the ISO 8601 format) and a deprecation reason:

```ruby
disable! date: "YYYY-MM-DD", because: :reason
```

The `date` parameter should be set to the date that the reason for disabling came into effect. If there is no clear date but the formula needs to be disabled, use today's date. If the `date` parameter is set to a date in the future, the formula will be deprecated until that date (on which the formula will become disabled).

The `because` parameter can be set to a preset reason (using a symbol) or a custom reason. See the [Deprecate and Disable Reasons](#deprecate-and-disable-reasons) section below for more details about the `because` parameter.

## Removal

> A formula should be removed if there is a serious issue with the formula or the formula has been disabled for a long period of time.

Reviewer 2: > for a long period of time.

After 1 year as stated above?

Author: I'm trying to avoid setting hard rules for third-party taps. I do have a note on the next line saying the 1 year rule for homebrew/core

Reviewer 4: @Rylan12 It may simplify this documentation to note that I've never seen anyone assume this documentation refers to third-party taps.

Author: Okay, fair point.

> A formula should be removed if there is a serious issue with the formula or the formula has been disabled for a long period of time.
> 
> A formula should be removed if it has been disabled for a long period of time, it has a non-open-source license, or there is another serious issue with the formula that makes it not compatible with homebrew/core.

Reviewer 5: ```suggestion
A formula should be removed if it has been disabled for a long period of time, it has a non-open-source license, or there is another serious issue with the formula that makes it not compatible with homebrew/core.
```

> A formula should be removed if it has been disabled for a long period of time, it has a non-open-source license, or there is another serious issue with the formula that makes it not compatible with homebrew/core.

Reviewer 3: For this you should also refer to the acceptance criteria for our definition of "not open source"

Reviewer: I actually prefer the terminology "non-free", as I think that corresponds more closely to what's actually not acceptable in homebrew-core. i.e. it doesn't matter if your source code is available if using/distributing it comes with an unduly restrictive license.

**Note: disabled formulae in homebrew/core will be automatically removed one year after their disable date**

## Deprecate and Disable Reasons

> When a formula is deprecated or disabled, a reason explaining the action should be provided.

Reviewer 4: ```suggestion
When a formula is deprecated or disabled, a reason explaining the action must be provided.
```

> There are two ways to indicate the reason. The preferred way is to use a pre-existing symbol to indicate the reason. The available symbols are:

Reviewer 4: Might be nice to link to this section from where the symbols are defined in the formula to encourage people adding any additional reasons to add to this doc.

Author: Is it better to link to the [`DeprecateDisable` module in the brew repo](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/deprecate_disable.rb) or the [`DeprecateDisable` module on rubydoc.brew.sh](https://rubydoc.brew.sh/DeprecateDisable.html)?

Reviewer 4: The prior, I think!

> - `:does_not_build`: the formulae that cannot be build from source

Reviewer 5: ```suggestion
- `:does_not_build`: the formulae that cannot be built from source
```

> - `:no_license`: the formulae does not have a license

Reviewer 3: Formulae is plural so..
```suggestion
- `:no_license`: the formulae do not have a license
```

- `:repo_archived`: the upstream repository has been archived
- `:repo_removed`: the upstream repository has been removed
- `:unmaintained`: the project appears to be abandoned
- `:unsupported`: Homebrew's application of the software is not supported by the upstream developers (e.g. upstream only supports macOS versions prior to 10.14)
- `:deprecated_upstream`: the project is deprecated upstream
- `:versioned_formula`: the formula is a versioned formula

These reasons can be specified by their symbols (the comments show the message that will be displayed to users):

```ruby
# Warning: <formula> has been deprecated because it is deprecated upstream!
deprecate! date: "2020-01-01", because: :deprecated_upstream
```

```ruby
# Error: <formula> has been disabled because it does not build!
disable! date: "2020-01-01", because: :does_not_build
```

If these pre-existing reasons do not fit, a custom reason can be specified. These reasons should be written to fit into the sentence `<formula> has been deprecated/disabled because it <reason>!`.

> Here are examples of a well-worded custom reason:

Reviewer 3: I'm not sure we have to teach people grammar in the homebrew documentation.

Author: I agree this may be too much handholding but we do have a few issues currently in homebrew/core so I don't think this part should totally be removed.

See [`liblwgeom`](https://github.com/Homebrew/homebrew-core/blob/3205250b454f50e27138b94f515aeda96dc73b06/Formula/liblwgeom.rb#L21):

```
Warning: liblwgeom has been deprecated because it liblwgeom headers are not installed anymore, use librttopo instead!
```

And [`exomizer`](https://github.com/Homebrew/homebrew-core/blob/3205250b454f50e27138b94f515aeda96dc73b06/Formula/exomizer.rb#L22):

```
Error: exomizer has been disabled because it custom license!
```

Reviewer 6: I think a formula author/maintainer would reasonably expect:
```
  ... because: "<reason>"
```
to produce the message:
```
  [This is happening] because <reason>.
```
rather than:
```
  [This is happening] because it <reason>!
```
In the case of `exomizer`, "because it custom license" is just _wrong_, but "because custom license" accords with modern usage, though it's anathema for grammar pedants. :laughing:

Author: @gromgit I agree that this makes sense. This is how I initially created it, but at some point, it was decided that since most of the reasons already started with "it", the "it" should be added by default. See https://github.com/Homebrew/brew/pull/8549#issuecomment-683924407 and https://github.com/Homebrew/homebrew-core/pull/60321#issuecomment-682642693. TBH I think it could be changed, now that we have the symbols available.

# Deprecating, Disabling, and Removing Formulae

There are many reasons why formulae may be deprecated, disabled, or removed. This document explains the differences between each method as well as explaining when one method should be used over another.

## Overview

This general rule of thumb can be followed:

- `deprecate!` should be used for formulae that _should_ no longer be used.
- `disable!` should be used for formulae that _cannot_ be used.

> - Formulae that have are not longer acceptable in homebrew/core or have been disabled for over a year should be removed.

Reviewer: ```suggestion
- Formulae that have are no longer acceptable in homebrew/core or have been disabled for over a year should be removed.
```

Author: Oops, should be:

```suggestion
- Formulae that are no longer acceptable in homebrew/core or have been disabled for over a year should be removed.
```

## Deprecation

If a user attempts to install a deprecated formula, they will be shown a warning message but the install will succeed.

> A formula should be deprecated to indicate to users that the formula should not be used and may be disabled in the future. Deprecated formulae should still be able to build from source and their bottles should continue to work.

Reviewer: ```suggestion
A formula should be deprecated to indicate to users that the formula should not be used and may be disabled in the future. Deprecated formulae should still build from source and their bottles should continue to work.
```

The most common reasons for deprecation are when the upstream project is deprecated, unmaintained, or archived.

To deprecate a formula, add a `deprecate!` call. This call should include a deprecation date (in the ISO 8601 format) and a deprecation reason:

```ruby
deprecate! date: "YYYY-MM-DD", because: :reason
```

The `date` parameter should be set to the date that the project or version became (or will become) deprecated. If there is no clear date but the formula needs to be deprecated, use today's date. If the `date` parameter is set to a date in the future, the formula will not become deprecated until that date. This can be useful if the upstream developers have indicated a date where the project or version will stop being supported.

> The `because` parameter can be set to a preset reason (using a symbol) or a custom reason. See the [Deprecate and Disable Reasons](#deprecate-and-disable-reasons) section below for more details about the `because` parameter.

Reviewer: ```suggestion
The `because` parameter can be a preset reason (using a symbol) or a custom reason. See the [Deprecate and Disable Reasons](#deprecate-and-disable-reasons) section below for more details about the `because` parameter.
```

## Disabling

If a user attempts to install a disabled formula, they will be shown an error message and the install will fail.

> A formula should be disabled to indicate to users that the formula cannot be used and will be removed in the future. Disabled formulae may no longer be able to build from source or have working bottles.

Reviewer: ```suggestion
A formula should be disabled to indicate to users that the formula cannot be used and will be removed in the future. Disabled formulae may no longer build from source or have working bottles.
```

> The most common reasons for disabling are when the formula cannot be built from source (meaning no bottles can be built), has been deprecated for a long time, the upstream repository has been removed, or the project has no license.

Reviewer: The parallelism in this sentence uses "when the formula":
1. cannot be built from source
2. has been deprecated for a long time
3. the upstream repository has been removed
4. the project has no license

so that doesn't quite work. You can fix it with the following, but I don't quite like it either.

```suggestion
The most common reasons for disabling are when the formula cannot be built from source (meaning no bottles can be built), the formula has been deprecated for a long time, the upstream repository has been removed, or the project has no license.
```

Author: Definitely better 👍

Author: Alternatively:

```suggestion
The most common reasons for disabling are when the formula cannot be built from source (meaning no bottles can be built), has been deprecated for a long time, had its upstream repository removed, or has no license.
```

or

```suggestion
The most common reasons for disabling are when the formula cannot be built from source (meaning no bottles can be built), has been deprecated for a long time, no longer has an upstream repository, or has no license.
```

Reviewer 6: Is there a reason to avoid bulleted lists here?

>The most common reasons for disabling are:
>* the formula cannot be built from source (meaning no bottles can be built)
>* it has been deprecated for a long time
>* no upstream repository
>* no license

Author: Nope, that's just how I wrote it. Makes sense to me to have it as a list. Good suggestion.

**Note: disabled formulae in homebrew/core will be automatically removed one year after their disable date**

To disable a formula, add a `disable!` call. This call should include a deprecation date (in the ISO 8601 format) and a deprecation reason:

```ruby
disable! date: "YYYY-MM-DD", because: :reason
```

The `date` parameter should be set to the date that the reason for disabling came into effect. If there is no clear date but the formula needs to be disabled, use today's date. If the `date` parameter is set to a date in the future, the formula will be deprecated until that date (on which the formula will become disabled).

> The `because` parameter can be set to a preset reason (using a symbol) or a custom reason. See the [Deprecate and Disable Reasons](#deprecate-and-disable-reasons) section below for more details about the `because` parameter.

Reviewer: ```suggestion
The `because` parameter can be a preset reason (using a symbol) or a custom reason. See the [Deprecate and Disable Reasons](#deprecate-and-disable-reasons) section below for more details about the `because` parameter.
```
